query-engine/core/src/telemetry/capturing/capturer.rs

```diff
@@ -156,70 +112,199 @@ impl SpanProcessor for Processor {
+                        } else {
+                            tracing::error!("storage should contain captures for (trace_id={:?})", trace_id);
+                            _ = sender.send(Storage::default());
+                        }
+                    }
+                    Ok(CaptureOp::GC(trace_id)) => {
+                        if store.remove(&trace_id).is_some() {
+                            tracing::trace!(
+                                "Gargabe collected after trace capturing started for trace_id={:?}.",
+                                trace_id
+                            );
+                        }
+                    }
+                    Err(_) => {
+                        tracing::error!("recv error in capture task");
+                        unreachable!("CAPTURE_TASK channel closed")
+                    }
+                }
+            }
+        });
 
-        let mut settings = Settings::default();
-        settings.ttl = one_ms;
+        sender
+    });
 
-        exporter.start_capturing(trace_id, settings).await;
-        let storage = exporter.storage.lock().unwrap();
-        assert!(storage.get(&trace_id).is_some());
-        drop(storage);
+    pub(super) fn span_data_processed(span_data: SpanData) -> Result<(), SendError<CaptureOp>> {
+        TASK.send(CaptureOp::SpanDataProcessed(span_data))
+    }
 
-        tokio::time::sleep(10 * one_ms).await;
+    pub(crate) async fn start_capturing(trace_id: TraceId, settings: Settings) -> Result<(), ()> {
+        let (sender, mut receiver) = tokio::sync::oneshot::channel::<()>();
+        TASK.send(CaptureOp::StartCapturing(trace_id, settings, sender))
+            .unwrap();
+
+        let mut interval = interval(Duration::from_millis(1));
+        let mut i = 0;
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    i+=1;
+                },
+                _msg = &mut receiver => {
+                    tracing::trace!("waited {}ms to receive signal from start capturing", i);
+                    return Ok(());
+                }
+            }
+        }
+    }
 
-        let storage = exporter.storage.lock().unwrap();
-        assert!(storage.get(&trace_id).is_none());
+    pub(crate) async fn fetch_captures_for_trace(
+        trace_id: TraceId,
+    ) -> Result<Storage, tokio::sync::oneshot::error::RecvError> {
+        let (sender, mut receiver) = oneshot::channel::<Storage>();
+        TASK.send(CaptureOp::FetchCaptures(trace_id, sender)).unwrap();
+        let mut interval = interval(Duration::from_millis(1));
+        let mut i = 0;
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    i+=1;
+                },
+                msg = &mut receiver => {
+                    tracing::trace!("waited {}ms to receive signal from fetch captures", i);
+                    return msg;
+                }
+            }
+        }
```

Comment:
There's an easier way to do this. This is pseudo-code but I am pretty sure this works with very little changes.

```rust
let (sender, mut receiver) = oneshot::channel::<Storage>();
TASK.send(CaptureOp::FetchCaptures(trace_id, sender)).unwrap();
let start = std::time::Instant::now();
let msg = receiver.await;
tracing::trace!("waited {}ms to receive signal from fetch captures", std::time::Instant::now().duration_since(start).as_millis());
msg
```

Reply:
Oh certainly! Yes :)